### PR TITLE
Add jit internally to a few slow routines in scatter_ops and linalg.

### PR DESCRIPTION
**Pull request by shoyer**
This makes these operations *much* faster when run outside of a jit context.

### CODE REVIEW

**Comment by mattjj** [comment]
Unfortunately this change caused test failures (Travis GH integration seems down at the moment) because these functions can take arguments that aren't supported by jit (e.g. Ellipsis):

https://travis-ci.org/google/jax/builds/555864962

**Comment by shoyer** [comment]
Would you rather try to support `Ellipsis` in `jax.jit` or wrap the lower level ops like `lax.scatter`? I think either would work here.

**Comment by mattjj** [comment]
Good question. I just remembered that this works:

```python
print(jit(lambda x: x)(None))
```

because [we register None as a pytree](https://github.com/google/jax/blob/76eda746bd28256fdba1b96036e8faa49088feea/jax/tree_util.py#L238). We could do the same with Ellipsis, adding this to tree_util.py (and also a rule for `slice` objects):

```python
register_pytree_node(type(Ellipsis), lambda z: ((), None), lambda _, xs: Ellipsis)
register_pytree_node(slice,
                     lambda s: ((s.start, s.stop, s.step), None),
                     lambda _, xs: slice(*xs))
```

However, that slice solution didn't pass our tests because of how we use the `slice.indices` method. This implementation does pass all tests, but assumes there aren't Tracers in the slice fields:

```python
register_pytree_node(slice,
                     lambda s: ((), (s.start, s.stop, s.step)),
                     lambda meta, _: slice(*meta))
```

If we can get the first solution for `slice` working, we'd be in business! Gotta look at it more.

**Comment by mattjj** [comment]
Ah the issue was that these added `@jit`s were then abstracting the contents of slices to the Shaped level, which would mean static slice indices were turned into dynamic slice indices. If we replace the use of `@jit` you added with `@partial(jit, static_argnums=(1,))` that avoids abstracting the indices and everything works!

Actually, I think it'd be better just to add a `@partial(jit, static_argnums=(1, 3))` decoration to `_scatter_update` rather than the wrapper functions.

This might be slightly more coarse than we need: we might be able to get more compilation cache hits if we just don't raise the abstraction level of (the contents of) `slice` instances. But that seems like more trouble than it's worth, since the point here is just to get these things under a jit.

@shoyer wdyt?

@dougalm any problem with registering `ellipsis` and `slice` as pytrees, like we do `NoneType`?

**Comment by mattjj** [comment]
I pushed 8c72e2f to a branch so you can take a look at the changes I proposed above.

File: jax/numpy/linalg.py [hunk 1]
```diff
@@ -222,6 +223,7 @@ def qr(a, mode="reduced"):
   return q, r
 
 
+@jit
```
Comment (hawkinsp) [inline jax/numpy/linalg.py]: FWIW, I added this `jit` decorator in PR #1144 for the same reason.

**Comment by shoyer** [comment]
@mattjj It looks like your branch handles this better than what I had here, so let's just abandon this PR in favor of that. It's also getting stale.

**Pull request closed without merging**